pkg/events/event_cache.go

```diff
@@ -0,0 +1,144 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package events
+
+import (
+	"sync"
+	"time"
+
+	"github.com/apache/incubator-yunikorn-core/pkg/log"
+)
+
+// TODO this should be configurable?
+const sleepTimeInterval = 10 * time.Millisecond
+const pushEventInterval = 2 * time.Second
+
+var once sync.Once
+var cache *EventCache
+
+type EventCache struct {
+	channel    EventChannel     // channelling input events
+	store      EventStore       // storing events
+	publishers []EventPublisher // publishing events to sinks
+	started    bool             // whether the service is started
+	stopped    bool             // whether the service is stopped
+
+	sync.Mutex
+}
+
+func GetEventCache() *EventCache {
+	once.Do(func(){
+		store := newEventStoreImpl()
+		pub := newShimPublisher(store)
+
+		cache = &EventCache{
+			channel:    newEventChannelImpl(),
+			store:      store,
+			publishers: []EventPublisher{pub},
+			started:    false,
+			stopped:    false,
+		}
+	})
+	return cache
+}
+
+// TODO consider thread pool
+func (ec *EventCache) StartService() {
+	ec.Lock()
+	defer ec.Unlock()
+
+	ec.started = true
+
+	// start main event processing thread
+	go ec.processEvent()
+
+	// start event publishers
+	for _, publisher := range ec.publishers {
+		publisher.StartService()
+	}
+}
+
+func (ec *EventCache) Stop() {
+	ec.Lock()
+	defer ec.Unlock()
+
+	if ec.started {
+		if ec.stopped {
+			panic("could not stop EventCache service")
+		}
+		ec.stopped = true
+
+		// stop event publishers
+		for _, publisher := range ec.publishers {
+			publisher.Stop()
+		}
+	}
+}
+
+func (ec *EventCache) AddPublisher(pub EventPublisher) {
+	ec.Lock()
+	defer ec.Unlock()
+
+	if ec.started  {
```

Comment:
I originally planned `Publisher` to be the part of the `EventCache`, but since we're moving towards a more flexible way to add/remove publishers I think we shouldn't forbid adding new publishers if we've already started the cache. Actually the state of the `EventCache` and the `EventPublishers` should not be bound to each other at all.

Modified the code accordingly.